slides/k8s/kubenet.md

```diff
@@ -82,22 +84,64 @@
 
 ---
 
+class: extra-details
+
 ## The Container Network Interface (CNI)
 
-- The CNI has a well-defined [specification](https://github.com/containernetworking/cni/blob/master/SPEC.md#network-configuration) for network plugins
+- Most Kubernetes clusters use CNI "plugins" to implement networking
+
+- When a pod is created, Kubernetes delegates the network setup to these plugins
 
-- When a pod is created, Kubernetes delegates the network setup to CNI plugins
+  (in can be a single plugin, or a combination of plugins, each doing one task)
 
-- Typically, a CNI plugin will:
+- Typically, CNI plugins will:
 
   - allocate an IP address (by calling an IPAM plugin)
 
   - add a network interface into the pod's network namespace
 
   - configure the interface as well as required routes etc.
 
-- Using multiple plugins can be done with "meta-plugins" like CNI-Genie or Multus
+---
+
+class: extra-details
+
+## Multiple moving parts
+
+- The "pod-to-pod network" or "pod network":
+
+  - provides communication between pods and nodes
+
+  - is generally implemented with CNI plugins
+
+- The "pod-to-service network":
+
+  - provides internal communication and load balancing
+
+  - is generally implemented with kube-proxy (or e.g. kube-router)
+
+- Network policies:
+
+  - provide firewalling and isolation
+
+  - can be bundled with the "pod network" or provided by another component
+
+---
+
+class: extra-details
+
+## Even more moving parts
+
+- Inbound traffic can be handled by multiple components:
+
+  - something like kube-proxy or kube-router (for NodePort services)
+
+  - load balancers (ideally, connected to the pod network)
+
+- It is possible to use multiple pod networks in parallel
+
+  (with "meta-plugins" like CNI-Genie or Multus)
 
-- Not all CNI plugins are equal
+- Some products can fill multiple roles
```

Comment:
Can you choose a less ambiguous word than "product" here?

Reply:
It's tricky! When I wrote my advanced Kubernetes course, I realized that I had been using the words "CNI plugin" incorrectly.

For instance, Weave is a product/solution implementing the pod Network, and ship a CNI plugin called `weave` to achieve so. It also comes with a Network Policy Controller. So it would be tempting to say, "I'm using the weave CNI plugin."

Now, kube-router is also a product/solution implementing the pod network, but it does so by leveraging the standard bridge CNI plugin. So it would be incorrect to say, "my CNI plugin is kube-router". It would be better to say, "I'm using kube-router to provide pod network (or pod connectivity)".

But kube-router can also be a replacement for kube-proxy!

Now Cilium can be used for pod connectivity, but also independently (with another pod network implementation) to provide advanced filtering (providing an implementation of the Network Policy Controller standard API group + their own API).

So ... rather than using incorrect terminology, I went for the vague "product" here but if you have a better idea, I'm buying it. Name your price! ;)

Reply:
I see what your'e saying. I would go with "solution" instead of "product" since "product" implies something specific that can be purchased from a specific vender, perhaps?

Reply:
Yes! That makes total sense. Thank you! I'll update right away.